### ISSUE TITLE
Compiler hangs

### ISSUE THREAD

**Issue by tindzk** [scala/pickling#247]
The following program causes the compiler to hang:

```scala
case class Segment(interps: Map[String, Double])
case class Sentence(segments: List[Segment])

BinaryPickleArray(Array()).unpickle[Map[String, List[Sentence]]]
```

See https://github.com/tindzk/pickling-test for a small sbt project.

There is also a sbt warning which I don't understand:

```
[warn] Binary version (0.9.2-SNAPSHOT) for dependency org.scala-lang#scala-pickling_2.11;0.9.2-SNAPSHOT
[warn]  in pickling-test#pickling-test_2.11;0.1-SNAPSHOT differs from Scala binary version in project (2.11).
```

**Comment by mgladdish** [comment]
And me (albeit without the sbt warning)

**Comment by eed3si9n** [comment]
I can't reproduce the compiler hanging using Pickler 0.10.0.

``` scala
scala-pickling> console
[info] Starting scala interpreter...
[info] 
Welcome to Scala version 2.11.4 (Java HotSpot(TM) 64-Bit Server VM, Java 1.7.0_51).
Type in expressions to have them evaluated.
Type :help for more information.

scala> import scala.pickling._, Defaults._, binary._
import scala.pickling._
import Defaults._
import binary._

scala> case class Segment(interps: Map[String, Double])
defined class Segment

scala> case class Sentence(segments: List[Segment])
defined class Sentence

scala> val pkl = Map.empty[String, List[Sentence]].pickle
warning: there was one deprecation warning; re-run with -deprecation for details
pkl: scala.pickling.binary.pickleFormat.PickleType = BinaryPickle([0,0,0,-120,115,99,97,108,97,46,99,111,108,108,101,99,116,105,111,110,46,105,109,109,117,116,97,98,108,101,46,77,97,112,91,106,97,118,97,46,108,97,110,103,46,83,116,114,105,110,103,44,115,99,97,108,97,46,99,111,108,108,101,99,116,105,111,110,46,105,109,109,117,116,97,98,108,101,46,76,105,115,116,91,36,108,105,110,101,49,50,46,36,114,101,97,100,46,36,105,119,46,36,105,119,46,36,105,119,46,36,105,119,46,36,105,119,46,36,105,119,46,36,105,119,46,36,105,119,46,83,101,110,116,101,110,99,101,93,93,0,0,0,0])

scala> pkl.unpickle[Map[String, List[Sentence]]]
warning: there was one deprecation warning; re-run with -deprecation for details
res0: Map[String,List[Sentence]] = Map()
```

The sbt warning issue is #142, which used happen because Pickling used to be published using org.scala-lang organization and sbt warns when it suspects Scala compiler and library versions do not match up. That's also now fixed in 0.10.0.
